csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -12,437 +13,366 @@ namespace ZeroC.Ice
+            if (reference.Communicator.TraceLevels.Location >= 1)
             {
-                var callback = new RequestCallback(reference, ttl, cb);
-                lock (_mutex)
+                if (endpoints != null && endpoints.Count > 0)
                 {
-                    if (!_response && _exception == null)
+                    if (cached)
                     {
-                        _callbacks.Add(callback);
-                        if (wellKnownRef != null)
+                        if (reference.IsWellKnown)
                         {
-                            // This request is to resolve the endpoints of a cached well-known object ref
-                            _wellKnownRefs.Add(wellKnownRef);
+                            Trace("found endpoints for well-known proxy in locator cache", reference, endpoints);
                         }
-                        if (!_sent)
+                        else
                         {
-                            _sent = true;
-                            Send();
+                            Trace("found endpoints for adapter in locator cache", reference, endpoints);
+                        }
+                    }
+                    else
+                    {
+                        if (reference.IsWellKnown)
+                        {
+                            Trace("retrieved endpoints for well-known proxy from locator, adding to locator cache",
+                                  reference, endpoints);
+                        }
+                        else
+                        {
+                            Trace("retrieved endpoints for adapter from locator, adding to locator cache",
+                                  reference, endpoints);
                         }
-                        return;
                     }
-                }
-
-                if (_response)
-                {
-                    callback.Response(LocatorInfo, _proxy);
                 }
                 else
                 {
-                    Debug.Assert(_exception != null);
-                    callback.Exception(_exception);
+                    Communicator communicator = reference.Communicator;
+                    var s = new System.Text.StringBuilder();
+                    s.Append("no endpoints configured for ");
+                    if (reference.AdapterId.Length > 0)
+                    {
+                        s.Append("adapter\n");
+                        s.Append("adapter = " + reference.AdapterId);
+                    }
+                    else
+                    {
+                        s.Append("well-known object\n");
+                        s.Append("well-known proxy = " + reference.ToString());
+                    }
+                    communicator.Logger.Trace(communicator.TraceLevels.LocationCat, s.ToString());
                 }
             }
+            return (endpoints ?? Array.Empty<Endpoint>(), cached);
```

Comment:
I think it might be simpler to switch to non-nullable `IReadOnlyList<Endpoint>` everywhere - no more `IReadOnlyList<Endpoint>?`.

Reply:
`_table.GetAdapterEndpoints()` needs to continue to return `IReadOnlyList<Endpoint>?` to figure out whether or not the endpoints are cached. I can fix `GetAdapterEndpointsAsync` to return `IReadOnlyList<Endpoint>` however and throw `NoEndpointException` if no endpoints are returned. Or do you have something else in mind?

Reply:
Why? I don't think an empty endpoint list is ever a valid endpoint list (in this context), so you could replace `endpoints != null` by `endpoints.Count > 0`.

See Reference.GetConnectionRequestHandlerAsync which calls these APIs with
```
if (endpoints == null || endpoints.Count == 0)
```

Reply:
Ok so your suggestion is to use an empty endpoint list instead of a nullable to indicate that we didn't resolve the endpoints yet. This should works yes.

Reply:
Yes. And on side-note, this should also be the preferred pattern in C++, with an empty vector (which costs nothing to allocate) instead of a `std::optional<std::vector<T>>`.

Reply:
I doubt we use optional in C++... we already use empty collections which are indeed cheap. We rely on null in C#/Java to avoid creating empty collections... We can rely on a static or better ImmutableXxx<T>.Empty if possible instead.